src/libs/ajax.js

```diff
@@ -306,23 +301,101 @@ export const Rawls = {
+      },
+
+      importBagit: bagitURL => {
+        return fetchOrchestration(
+          `/api/workspaces/${namespace}/${name}/importBagit`,
+          _.mergeAll([authOpts(), jsonBody({ bagitURL, format: 'TSV' }), { method: 'POST' }])
+        )
       }
     }
+  }
+}
+
+
+export const Buckets = {
+  getObject: async (bucket, object, namespace) => {
+    return fetchBuckets(`storage/v1/b/${bucket}/o/${encodeURIComponent(object)}`,
+      authOpts(await User.token(namespace))).then(
+      res => res.json()
+    )
   },
 
-  methodConfigInputsOutputs: async loadedConfig => {
+  getObjectPreview: async (bucket, object, namespace, previewFull = false) => {
+    return fetchBuckets(`storage/v1/b/${bucket}/o/${encodeURIComponent(object)}?alt=media`,
+      _.merge(authOpts(await User.token(namespace)), previewFull ? {} : { headers: { range: 'bytes=20000' } }))
+  },
+
+  listNotebooks: async (namespace, name) => {
+    const res = await fetchBuckets(
+      `storage/v1/b/${name}/o?prefix=notebooks/`,
+      authOpts(await User.token(namespace))
+    )
+    const { items } = await res.json()
+    return _.filter(({ name }) => name.endsWith('.ipynb'), items)
+  },
+
+  notebook: (namespace, bucket, name) => {
+    const bucketUrl = `storage/v1/b/${bucket}/o`
+
+    const copy = async newName => {
+      return fetchBuckets(
+        `${bucketUrl}/${nbName(name)}/copyTo/b/${bucket}/o/${nbName(newName)}`,
+        _.merge(authOpts(await User.token(namespace)), { method: 'POST' })
+      )
+    }
+    const doDelete = async () => {
+      return fetchBuckets(
+        `${bucketUrl}/${nbName(name)}`,
+        _.merge(authOpts(await User.token(namespace)), { method: 'DELETE' })
+      )
+    }
+    return {
+      copy,
+
+      create: async contents => {
+        return fetchBuckets(
+          `upload/${bucketUrl}?uploadType=media&name=${nbName(name)}`,
+          _.merge(authOpts(await User.token(namespace)), {
+            method: 'POST', body: JSON.stringify(contents),
+            headers: { 'Content-Type': 'application/x-ipynb+json' }
+          })
+        )
+      },
+
+      delete: doDelete,
+
+      rename: async newName => {
+        await copy(newName)
+        return doDelete()
+      }
+    }
+  }
+}
+
+
+export const Methods = {
+  configInputsOutputs: async loadedConfig => {
     const res = await fetchRawls('methodconfigs/inputsOutputs',
       _.mergeAll([authOpts(), jsonBody(loadedConfig.methodRepoMethod), { method: 'POST' }]))
     return res.json()
-  }
-}
+  },
 
+  method: (namespace, name, snapshotId) => {
+    const root = `methods/${namespace}/${name}/${snapshotId}`
 
-const fetchLeo = async (path, ...args) => {
-  return fetchOk(`${await Config.getLeoUrlRoot()}/${path}`, ...args)
+    return {
+      get: async () => {
+        const res = await fetchAgora(root, authOpts())
+        return res.json()
+      }
+    }
+  }
 }
 
-export const Leo = {
+
+export const Jupyter = {
```

Comment:
TOL: My gut instinct is usually to avoid creating taxonomies entirely, since they tend to be reflective of the minds of the developers, rather than any objective truth of the underlying domain. However, I also know they can be helpful organizing principles, and the risk in this case is minimal, so I think it's fine.

Reply:
In this case, I stuck them together because they are _Jupyter clusters_ and _Jupyter notebooks_ when I refer to them in my head.